features/dnsresolver/netty/src/main/java/org/opennms/netmgt/dnsresolver/netty/NettyDnsResolver.java

```diff
@@ -0,0 +1,267 @@
+import io.netty.resolver.dns.DnsServerAddressStreamProvider;
+import io.netty.resolver.dns.DnsServerAddressStreamProviders;
+import io.netty.resolver.dns.SequentialDnsServerAddressStreamProvider;
+import io.netty.util.internal.SocketUtils;
+
+/**
+ * Asynchronous DNS resolution using Netty.
+ *
+ * Creates multiple resolvers (aka contexts) (defaults to 2*num cores) against which the queries
+ * are randomized in order to improve performance.
+ *
+ * Uses a circuit breaker in order to ensure that callers do not continue to be bogged down
+ * if resolution fails.
+ *
+ * @author jwhite
+ */
+public class NettyDnsResolver implements DnsResolver {
+    private static final Logger LOG = LoggerFactory.getLogger(NettyDnsResolver.class);
+
+    public static final String CIRCUIT_BREAKER_STATE_CHANGE_EVENT_UEI = "uei.opennms.org/circuitBreaker/stateChange";
+
+    private final EventForwarder eventForwarder;
+    private final Timer lookupTimer;
+    private final Meter lookupsSuccessful;
+    private final Meter lookupsFailed;
+    private final Meter lookupsRejectedByCircuitBreaker;
+
+    private int numContexts = 0;
+    private String nameservers = null;
+    private long queryTimeoutMillis = TimeUnit.SECONDS.toMillis(5);
+    private int minTtlSeconds = -1;
+    private int maxTtlSeconds = -1;
+    private int negativeTtlSeconds = -1;
+
+    private List<NettyResolverContext> contexts;
+    private Iterator<NettyResolverContext> iterator;
+    private DnsCache cache;
+
+    private CircuitBreaker circuitBreaker;
+
+    public NettyDnsResolver(EventForwarder eventForwarder, MetricRegistry metrics) {
+        this.eventForwarder = Objects.requireNonNull(eventForwarder);
+        lookupTimer = metrics.timer("lookups");
+        lookupsSuccessful = metrics.meter("lookupsSuccessful");
+        lookupsFailed = metrics.meter("lookupsFailed");
+        lookupsRejectedByCircuitBreaker = metrics.meter("lookupsRejectedByCircuitBreaker");
+        // It would be nice to expose cache statistics too, but Netty's cache doesn't currently
+        // make any statistics available - see https://github.com/netty/netty/issues/9412
+    }
+
+    public void init() {
+        numContexts = Math.max(0, numContexts);
+        if (numContexts == 0) {
+            numContexts = Runtime.getRuntime().availableProcessors() * 2;
+        }
+        LOG.debug("Initializing Netty resolver with {} contexts and resolvers: {}", numContexts);
+
+
+        contexts = new ArrayList<>(numContexts);
+
+        // Initialize the cache with the given TTL settings - use Netty's default if the configured values
+        // are less than 0
+        final DefaultDnsCache cacheWithDefaults = new DefaultDnsCache();
+        cache = new DefaultDnsCache(minTtlSeconds < 0 ? cacheWithDefaults.minTtl() : minTtlSeconds,
+                maxTtlSeconds < 0 ? cacheWithDefaults.maxTtl() : maxTtlSeconds,
+                negativeTtlSeconds < 0 ? cacheWithDefaults.negativeTtl(), negativeTtlSeconds);
```

Comment:
This does not compile. Must be a colon: "negativeTtlSeconds < 0 ? cacheWithDefaults.negativeTtl() : negativeTtlSeconds"

Reply:
Fixed.